File: examples/src/main/java/recommendations/tutorial4/Serving.java
package io.prediction.examples.java.recommendations.tutorial4;

import io.prediction.controller.java.LJavaServing;
import io.prediction.controller.java.EmptyParams;
import java.lang.Iterable;

public class Serving extends LJavaServing<EmptyParams, Query, Float> {
  public Serving() {}

  public Float serve(Query query, Iterable<Float> predictions) {
    float sum = 0.0f;
    int count = 0;
    
    for (Float v: predictions) {
      if (!v.isNaN()) {
        sum += v;
        count += 1;
      }
    }
    return (count == 0) ? Float.NaN : sum / count;
  }
}

File: examples/scala-parallel-recommendation-advanced/src/main/scala/Serving.scala
package org.examples.recommendation

import io.prediction.controller._

class Serving
  extends Serving[EmptyServingParams, Query, PredictedResult] {

  override
  def serve(query: Query,
    predictions: Seq[PredictedResult]): PredictedResult = {
    predictions.head
  }
}
